Data/ByteString/Internal/Type.hs

```diff
@@ -530,6 +532,104 @@ packUptoLenChars len cs0 =
           go !p (c:cs)          = pokeFp p (c2w c) >> go (p `plusForeignPtr` 1) cs
       in go p0 cs0
 
+#if MIN_VERSION_template_haskell(2,17,0)
+type THLift a = forall m. (MonadFail m, TH.Quote m) => TH.Code m a
+
+liftTyped :: forall a m. (MonadFail m, TH.Quote m, TH.Lift a) => a -> TH.Code m a
+liftTyped = TH.liftTyped
+
+liftCode :: forall a m. (MonadFail m, TH.Quote m) => m (TH.TExp a) -> TH.Code m a
+liftCode = TH.liftCode
+#else
+type THLift a = TH.Q (TH.TExp a)
+
+liftTyped :: forall a. TH.Lift a => a -> TH.Q (TH.TExp a)
+liftTyped = TH.unsafeTExpCoerce . TH.lift
+
+liftCode :: forall a. TH.Q TH.Exp -> TH.Q (TH.TExp a)
+liftCode = TH.unsafeTExpCoerce
+#endif
+
+data S2W = Octets {-# UNPACK #-} !Int [Word8]
+           -- ^ Decoded some octets (<= 0xFF)
+         | Hichar {-# UNPACK #-} !Int {-# UNPACK #-} !Word
+           -- ^ Found a high char (> 0xFF)
+
+data H2W = Hex {-# UNPACK #-} !Int [Word8]
+           -- ^ Decoded some full bytes (nibble pairs)
+         | Odd {-# UNPACK #-} !Int {-# UNPACK #-} !Word [Word8]
+           -- ^ Decoded a nibble plus some full bytes
+         | Bad {-# UNPACK #-} !Int {-# UNPACK #-} !Word
+           -- ^ Found a non hex-digit character
+
+-- | Template Haskell splice to convert string constants to compile-time
+-- ByteString literals.  Unlike the 'IsString' instance, the input string
+-- is validated to ensure that each character is a valid /octet/, i.e. is
+-- at most @0xFF@ (255).
+--
+-- Example:
+--
+-- > :set -XTemplateHaskell
+-- > ehloCmd :: ByteString
+-- > ehloCmd = $$(literalFromChar8 "EHLO")
+--
+literalFromChar8 :: String -> THLift ByteString
+literalFromChar8 "" = [||empty||]
+literalFromChar8 s = case foldr' op (Octets 0 []) s of
+    Octets n ws -> liftTyped (unsafePackLenBytes n ws)
+    Hichar i w  -> liftCode $ fail $ "non-octet character '\\" ++
+        show w ++ "' at offset: " ++ show i
+  where
+    op :: Char -> S2W -> S2W
+    op (fromIntegral . fromEnum -> !(w :: Word)) acc
+        | w <= 0xff = case acc of
+            Octets i ws -> Octets (i + 1) (fromIntegral w : ws)
```

Comment:
`(i+1)` can overflow.

Reply:
That would require that the input string length be greater than `maxBound @Int`, and the consequence is just an incorrect report of the problem offset.  I suspect this does not warrant work-arounds.